crates/rust-analyzer/src/main_loop.rs

```diff
@@ -484,56 +500,64 @@ fn on_notification(
     };
     let not = match notification_cast::<lsp_types::notification::DidOpenTextDocument>(not) {
         Ok(params) => {
-            let uri = params.text_document.uri;
-            let path = uri.to_file_path().map_err(|()| format!("invalid uri: {}", uri))?;
-            if let Some(file_id) =
-                state.vfs.write().add_file_overlay(&path, params.text_document.text)
-            {
-                loop_state.subscriptions.add_sub(FileId(file_id.0));
+            if let Ok(path) = from_proto::vfs_path(&params.text_document.uri) {
+                if !loop_state.mem_docs.insert(path.clone()) {
+                    log::error!("duplicate DidOpenTextDocument: {}", path)
+                }
+                global_state
+                    .vfs
+                    .write()
+                    .0
+                    .set_file_contents(path, Some(params.text_document.text.into_bytes()));
             }
             return Ok(());
         }
         Err(not) => not,
     };
     let not = match notification_cast::<lsp_types::notification::DidChangeTextDocument>(not) {
         Ok(params) => {
-            let DidChangeTextDocumentParams { text_document, content_changes } = params;
-            let world = state.snapshot();
-            let file_id = from_proto::file_id(&world, &text_document.uri)?;
-            let line_index = world.analysis().file_line_index(file_id)?;
```

Comment:
Do you think it's not worth keeping the `Cow`?

Reply:
Hm, I thought that I can't re-use the Cow, as the file might not be in the Analysis, but looks like I was wrong? Yeah, this probably should go back to Cow, though, I don't think it matter that much.